promql: add step hint from subqueries

anukul opened the pull request:
Fixes #7630

I think we can choose the GCD of the steps specified by all subqueries.
For example, in the query:

```
sum_over_time(rate(node_cpu_seconds_total[3m:16s])[5m:20s])
```

`sum_over_time` will use the value of `rate(node_cpu_seconds_total[3m:16s])` calculated at every `20s` step,
while `rate` will use the values of `node_cpu_seconds_total` at every `16s` step.

So, we can fetch data points from the time series at every `4s` step and fulfil the requirements of both subqueries.

anukul commented:
@bwplotka Hi! If you have any suggestions for further improvements, I'm happy to add those too.
Otherwise please let me know if I should remove the TODOs in code.

brian-brazil commented:
> I think we can choose the GCD of the steps specified by all subqueries.

Why do you think this?

> sum_over_time(rate(node_cpu_seconds_total[3m:16s])[5m:20s])

This is not a sane query, and there's no need to return more than 16s here as the inner step is independent of everything else.

anukul commented:
>> I think we can choose the GCD of the steps specified by all subqueries.

> Why do you think this?

Sorry. I had assumed that the step size of a nested range-vector should exactly divide the step size of all parent subqueries, since they access its return value in intervals of their own step size.

From your explanation, it seems like my understanding of this evaluation logic is incorrect, and parent subqueries will instead use the last value available in the range-vector at every step interval.

So.. is it safe to say that a subquery should only affect the step hint when its expression is a vector selector (and not a function call)?

brian-brazil commented:
> parent subqueries will instead use the last value available in the range-vector at every step interval.

They will follow standard range/instance vector evaluation semantics.

> So.. is it safe to say that a subquery should only affect the step hint when its expression is a vector selector (and not a function call)?

That's not the way to look at things. The question is what optimisation the other end is trying to do and what information does it need to correctly perform that optimisation?

anukul commented:
Ok, I'll try to answer this.

> what optimisation the other end is trying to do

When populating the time series for a vector selector, we want to use a resolution that gives us values for timestamps at all steps of the subquery that contains the vector selector.

> what information does it need to correctly perform that optimisation?

The step size of the subquery that contains the vector selector.
Additionally, for range queries, we should use `gcd(range_query_interval, subquery_step_size)` so we fetch data points for all timestamps under consideration.

/**
e.g.
```
query: max_over_time(foo[16s:16s])
start: t
end: t + 10s
interval: 10s
```

evaluation 1: `t`, `t - 16s`
evaluation 2: `t + 10s`, `t - 6s`

If we use a step size of `16s` for resolution, both evaluations will use the same data points, which is not desirable.
**/

Is this correct? I hope I'm not misunderstanding the evaluation semantics again.

brian-brazil commented:
I still fail to see what gcd has to do with any of this. I'm also not following your example, which is not a sane query and is also underspecified.

Either the subquery step is sufficient or it isn't.

anukul commented:
Based on the discussion in IRC, I have updated the PR accordingly.

I have also updated one of the test cases to clarify that in the case of range queries, the subquery step size is now the step hint. We were previously using the range query interval as the step hint, but it is now ignored in such cases.

This was also the source of my confusion, thanks for clarifying!

brian-brazil commented on promql/engine.go:
```diff
@@ -651,6 +654,10 @@ func (ng *Engine) populateSeries(querier storage.Querier, s *parser.EvalStmt) {
 			}
 
+			if evalStep != 0 {
+				hints.Step = durationMilliseconds(evalStep)
```
We shouldn't be changing the meaning of existing hints, rather add a new hint.

brian-brazil commented on promql/engine.go:
```diff
@@ -622,14 +622,18 @@ func (ng *Engine) populateSeries(querier storage.Querier, s *parser.EvalStmt) {
 	var evalRange time.Duration
+	// Whenever a SubqueryExpr is evaluated, subqueryStep is set to the corresponding step.
+	// The evaluation of the VectorSelector inside then uses this.
+	var subqueryStep time.Duration
```
This is a local over the whole expression, so if you had somethging like `max_over_time(a[1s:1s]) + a` then we'd end up incorrectly sending a subquery step with the 2nd selector.

brian-brazil reviewed (commented):
You also need to plumb this through to the proto, which is where the hints are actually exposed.

brian-brazil commented on promql/engine_test.go:
```diff
@@ -355,13 +355,13 @@ func TestSelectHintsSetCorrectly(t *testing.T) {
 		expected: []*storage.SelectHints{
-			{Start: 95000, End: 120000, Func: "sum", By: true},
-			{Start: 95000, End: 120000, Func: "max", By: true},
+			{Start: 95000, End: 120000, Func: "sum", By: true, SubqueryStep: 5000},
```
This is wrong.

anukul replied:
What is the expected value?

brian-brazil replied:
None, as this isn't part of a subquery.
